compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -31,7 +31,6 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         error: TypeError<'tcx>,
     ) {
         self.annotate_expected_due_to_let_ty(err, expr, error);
-        self.suggest_box_deref(err, expr, expected, expr_ty);
         self.suggest_compatible_variants(err, expr, expected, expr_ty);
         self.suggest_deref_ref_or_into(err, expr, expected, expr_ty, expected_ty_expr);
```

Comment:
I think you just need to move this line up one (in a separate commit, please). If that doesn't have too much fallout, then I think it would solve my concern about the variant suggestion coming first.